### PR TITLE
devops: Install pre-commit into devcontainer

### PR DESCRIPTION
**Pull request by max-sixty**
Does https://github.com/PRQL/prql/issues/3149#issuecomment-1676590241 automatically

### CODE REVIEW

**Review by eitsupi — CHANGES_REQUESTED** [review]
I think the installation command is not optimized. There is a cache issue.

And it is definitely more appropriate to do it in postCreateCommand instead of on the Dockerfile.

File: .devcontainer/base-image/Dockerfile [hunk 1]
```diff
@@ -12,6 +12,8 @@ RUN apt-get update \
 COPY Taskfile.yml /tmp/Taskfile.yml
 USER vscode
 
-RUN task -t /tmp/Taskfile.yml install-cargo-tools-binstall
+RUN task -t /tmp/Taskfile.yml install-cargo-tools-binstall \
+    && task install-pre-commit \
+    && pre-commit install-hooks
```
Comment (eitsupi) [inline .devcontainer/base-image/Dockerfile:17]: Does this work?

**Review by eitsupi — CHANGES_REQUESTED** [review]

File: .devcontainer/base-image/Dockerfile (hunk 1 above)
Comment (eitsupi) [reply .devcontainer/base-image/Dockerfile:17]: `pre-commit install-hooks` would make the commit fail when committing outside the container.

Comment (max-sixty) [reply .devcontainer/base-image/Dockerfile:17]: > `pre-commit install-hooks` would make the commit fail when committing outside the container.

Hmm, why?

NB: this is different from `pre-commit install` — it just installs the current hooks, it doesn't have them run on each commit...

**Comment by max-sixty** [comment]
> There is a cache issue. And it is definitely more appropriate to do it in postCreateCommand instead of on the Dockerfile.

I was thinking it would be better for the cache to do them here, so that it's cached? 

Or the versions will be out of date and so the hooks will often need to be reinstalled?

(tbc you know this better, I defer to you...)

File: .devcontainer/base-image/Dockerfile (hunk 1 above)
Comment (eitsupi) [reply .devcontainer/base-image/Dockerfile:17]: > NB: this is different from `pre-commit install` — it just installs the current hooks, it doesn't have them run on each commit...

Ah, my mistake. Thanks for pointing that out.

**Comment by eitsupi** [comment]
The installation of Rust tools takes so long that we have no choice but to install them on the Dockerfile, but I think it is far more flexible for users to install things that take an extremely long time to install, rather than including them in the pre-build image.

> I was thinking it would be better for the cache to do them here, so that it's cached?

Sorry I couldn't explain it adequately from my smart phone. What I meant was that the `pip install` command bakes the cache file into the container. 

https://github.com/PRQL/prql/blob/34507c322da0bd69bc5983c56b156d8ad4f83310/Taskfile.yml#L144-L145

Here it is common to use the `--no-cache-dir` option.

**Comment by eitsupi** [comment]
I looked at it again and thought that I should really rewrite the contents of the Taskfile.....

We should not be forced to install something as huge as a language runtime via Homebrew....

**Comment by eitsupi** [comment]
Closing in favor of #3344

File: .devcontainer/base-image/Dockerfile (hunk 1 above)
Comment (eitsupi) [reply .devcontainer/base-image/Dockerfile:17]: As an addendum, I think `task install-pre-commit` will fail because it cannot find the Taskfile.

**Comment by max-sixty** [comment]
> Sorry I couldn't explain it adequately from my smart phone. What I meant was that the `pip install` command bakes the cache file into the container.

Ah yes, very good point...

> We should not be forced to install something as huge as a language runtime via Homebrew....

I agree it's not great. What's the alternative? Maybe we stop & print some instructions if python isn't installed, in the same way we do for VS Code extensions?

**Comment by max-sixty** [comment]
> The installation of Rust tools takes so long that we have no choice but to install them on the Dockerfile, but I think it is far more flexible for users to install things that take an extremely long time to install, rather than including them in the pre-build image.

To confirm, do you mean "for users to install things that _don't_ take an extremely long time to install"?

**Comment by eitsupi** [comment]
> To confirm, do you mean "for users to install things that _don't_ take an extremely long time to install"?

I means that the installation is done via user side by `postCreateCommand` etc. using the definition files that are included in the repository at the time.

**Comment by eitsupi** [comment]
> I agree it's not great. What's the alternative? Maybe we stop & print some instructions if python isn't installed, in the same way we do for VS Code extensions?

I think that would be better.

For example, I think the current command forces Python3 installation via brew, even for those who prefer to build and install Python from source. This is quite surprising.
It is far better to fail to build this project than to be forced to install something.

**Pull request closed without merging**